api/src/main/java/org/openmrs/module/pihcore/event/PatientUpdateEventConsumer.java

```diff
@@ -0,0 +1,209 @@
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Consumes patient-related events
+ * For performance reasons on larger databases, this does not use a Debezium-generated snapshot, but rather
+ * computes a custom snapshot that is then updated as needed from change events.
+ * When used, this should be configured with "snapshot.mode = schema_only"
+ */
+public class PatientUpdateEventConsumer implements EventConsumer {
+
+    private static final Logger log = LogManager.getLogger(PatientUpdateEventConsumer.class);
+
+    private final DbEventSourceConfig config;
+    private final Database database;
+    private final Map<String, String> patientKeys = new LinkedHashMap<>();
+    private Rocks statusDb;
+    private boolean snapshotInitialized = false;
+
+    public PatientUpdateEventConsumer(DbEventSourceConfig config) {
+        this.config = config;
+        this.database = config.getContext().getDatabase();
+        patientKeys.put("patient_id", "patient");
+        patientKeys.put("person_id", "person");
+        patientKeys.put("person_a", "person");
+        patientKeys.put("person_b", "person");
+        patientKeys.put("order_id", "orders");
+        patientKeys.put("patient_program_id", "patient_program");
+        patientKeys.put("encounter_id", "encounter");
+        patientKeys.put("visit_id", "visit");
+        patientKeys.put("allergy_id", "allergy");
+        patientKeys.put("appointment_id", "appointmentscheduling_appointment");
+    }
+
+    @Override
+    public void startup() {
+        try {
+            statusDb = new Rocks(new File(config.getContext().getModuleDataDir(), "status.db"));
+            snapshotInitialized = BooleanUtils.isTrue(statusDb.get("snapshotInitialized"));
+        }
+        catch (Exception e) {
+            throw new RuntimeException("An error occurred starting up", e);
+        }
+    }
+
+    @Override
+    public void shutdown() {
+        statusDb.close();
+    }
+
+    @Override
+    public void accept(DbEvent event) {
+        if (!snapshotInitialized) {
+            performInitialSnapshot();
+            statusDb.put("snapshotInitialized", Boolean.TRUE);
+            snapshotInitialized = true;
+        }
+        Set<Integer> patientIds = getPatientIdsForEvent(event);
+        if (patientIds.isEmpty()) {
+            log.warn("Not handling event as not mapped to patient: " + event);
+        }
+        else {
+            for (Integer patientId : patientIds) {
+                handlePatientEvent(patientId, event);
+            }
+        }
+    }
+
+    /**
+     * For the given event, returns the associated patient ids
+     * @param event the event to retrieve the related patient ids
+     * @return the patient ids associated with the given event
+     */
+    public Set<Integer> getPatientIdsForEvent(DbEvent event) {
+        Set<Integer> ret = new HashSet<>();
```

Comment:
The goal of this method is to get the patient_id associated with a given event.  This is easy when "patient_id" is a column on the table.  Otherwise, we need to determine how to retrieve the patient id for the given table and column(s).  This - as far as I've figured - covers all of our use cases.  I have other solutions I've played around with that are more generic.  We can evolve this as we see fit